### PR TITLE
Length of longest substring

### PR DESCRIPTION
**Pull request by Urz1**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_length_of_longest_substring.py
Comment (ciiyaa) [inline solutions/tests/test_length_of_longest_substring.py:1]: Missing a header.
For the unittests, each test should have only one assertion and a clear docstring describing a single case.

File: solutions/tests/test_length_of_longest_substring.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import unittest
+from solutions.length_of_longest_substring import Solution
+
+
+class TestLengthOfLongestSubstring(unittest.TestCase):
+    """Unit tests for lengthOfLongestSubstring function."""
```
Comment (ciiyaa) [inline solutions/tests/test_length_of_longest_substring.py:6]: Should match the name of the function
```suggestion
    """Unit tests for length_of_longest_substring function."""
```

**Review by ciiyaa — CHANGES_REQUESTED** [review]
Reviewed.

**Comment by Urz1** [comment]
@ciiyaa Thanks for your valuable insights i have tried to correct every crucial suggestion check it out.

File: solutions/length_of_longest_substring.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      longest substring without repeating characters.
+
+Challenge:
+    This problem is sourced from various coding platforms, including LeetCode.
+
+Example Usage:
+    >>> from longest_substring import Solution
+    >>> s = "abcabcbb"
+    >>> Solution().length_of_longest_substring(s)
+    3
```
Comment (ciiyaa) [inline solutions/length_of_longest_substring.py:21]: This part raises **ModuleNotFoundError**, imports within the same file are unnecessary and can be removed

**Review by ciiyaa — CHANGES_REQUESTED** [review]
Thank you for addressing the previous feedback and making the updates! While the code could be implemented without the _Solutions_ class, I understand this is a matter of personal preference, and the current approach works fine. I just have one more suggestion noted in this review, kindly check it.

File: solutions/length_of_longest_substring.py (hunk 2 above)
Comment (Urz1) [reply solutions/length_of_longest_substring.py:21]: I really appreciate your suggestion. But here it's inside a comment and won't affect the program at all it's completely ignored while execusion so no worries it just tehre for illustration puprose.

File: solutions/tests/test_length_of_longest_substring.py (hunk 1 above)
Comment (Urz1) [reply solutions/tests/test_length_of_longest_substring.py:13]: noted!

**Review by Urz1 — COMMENTED** [review]
Thanks ava as always i have added the incorporated the appropriate suggsetion and improved the code. But It seems there are the same issue raised in the last review. I made the necessary changes in commit in the last commit, where I . Could you confirm if this addresses your concerns? If not, I’d appreciate more guidance on what might still need improvement. i,e for outdated cases.

**Review by ciiyaa — APPROVED** [review]
Thank you so much for addressing the feedback! great work!

**Comment by Urz1** [comment]
Thank you so much for all your invaluable insights!
